File: src/main/kotlin/com/ederfmatos/reactive/java/SyncOne.java
package com.ederfmatos.reactive.java;

import com.fasterxml.jackson.databind.annotation.JsonSerialize;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

@JsonSerialize(using = SyncOneSerializer.class)
public class SyncOne<T> implements One<T> {

    final T value;

    private SyncOne(T value) {
        this.value = value;
    }

    @Override
    public <R> One<R> map(Function<T, R> transformer) {
        if (isEmpty()) return empty();
        return new SyncOne<>(transformer.apply(value));
    }

    @Override
    public <R> One<R> flatMap(Function<T, One<R>> transformer) {
        if (isEmpty()) return empty();
        return transformer.apply(value);
    }

    @Override
    public <R> One<R> then(Supplier<One<R>> transformer) {
        return transformer.get();
    }

    @Override
    public <R> One<R> thenConsume(Supplier<R> transformer) {
        return new SyncOne<>(transformer.get());
    }

    @Override
    public One<T> ifEmpty(Supplier<One<T>> supplier) {
        if (isEmpty()) return supplier.get();
        return this;
    }

    @Override
    public One<T> ifEmptyReturn(Supplier<T> supplier) {
        if (isEmpty()) return new SyncOne<>(supplier.get());
        return this;
    }

    @Override
    public One<T> filter(Predicate<T> filter) {
        if (isEmpty() || filter.test(value)) return this;
        return empty();
    }

    @Override
    public One<T> filterNot(Predicate<T> filter) {
        if (isEmpty() || !filter.test(value)) return this;
        return empty();
    }

    @Override
    public One<T> onNext(Consumer<T> consumer) {
        if (isNotEmpty()) consumer.accept(value);
        return this;
    }

    @Override
    public One<T> onError(Consumer<Throwable> consumer) {
        if (value instanceof Throwable throwable) consumer.accept(throwable);
        return this;
    }

    @Override
    public One<T> onEmpty(Runnable runnable) {
        if (isEmpty()) runnable.run();
        return this;
    }

    @Override
    public One<T> onSuccess(Consumer<T> consumer) {
        if (isNotEmpty()) consumer.accept(value);
        return this;
    }

    @Override
    public One<T> onComplete(Runnable runnable) {
        if (isNotEmpty()) runnable.run();
        return this;
    }

    private boolean isEmpty() {
        return value == null;
    }

    private boolean isNotEmpty() {
        return value != null;
    }

    public static <T> One<T> empty() {
        return new SyncOne<>(null);
    }

    public static <T> One<T> from(T value) {
        return new SyncOne<>(value);
    }
}

File: src/main/kotlin/com/ederfmatos/reactive/kotlin/SyncOne.kt
package com.ederfmatos.reactive.kotlin

import com.fasterxml.jackson.databind.annotation.JsonSerialize

@JsonSerialize(using = SyncOneSerializer::class)
class SyncOne<T : Any>(val value: T?) : One<T> {

    override fun <R : Any> map(block: (T) -> R): One<R> {
        if (value == null) return empty()
        return block(value).let(::SyncOne)
    }

    override fun <R : Any> flatMap(block: (T) -> One<R>): One<R> {
        if (value == null) return empty()
        return block(value)
    }

    override fun ifEmpty(block: () -> One<T>): One<T> {
        if (value == null) return block()
        return this
    }

    override fun ifEmptyReturn(block: () -> T): One<T> {
        if (value == null) return block().let(::SyncOne)
        return this
    }

    override fun filter(block: (T) -> Boolean): One<T> {
        return if (value == null || block.invoke(value)) this
        else empty()
    }

    override fun filterNot(block: (T) -> Boolean): One<T> {
        return if (value == null || !block.invoke(value)) this
        else empty()
    }

    override fun onNext(block: (T) -> Unit): One<T> {
        if (value == null) return empty()
        block(value)
        return this
    }

    override fun onError(block: (Throwable) -> Unit): One<T> {
        if (value == null) return empty()
        if (value is Throwable) block(value)
        return this
    }

    override fun onEmpty(block: () -> Unit): One<T> {
        if (value == null) block()
        return this
    }

    override fun onComplete(block: () -> Unit): One<T> {
        block()
        return this
    }

    override fun onSuccess(block: (T) -> Unit): One<T> {
        if (value == null) return empty()
        block(value)
        return this
    }

    companion object {
        fun <T : Any> empty(): SyncOne<T> {
            return SyncOne(null)
        }

        fun <T : Any> from(value: T): One<T> {
            return SyncOne(value)
        }
    }
}